Cilium envoy crd

jrajahalme opened the pull request:
**Note:** Generic fixes in this series are to be reviewed in their own PRs. This PR will be rebased as these PRs are merged:
- [x] https://github.com/cilium/cilium/pull/17591
- [x] https://github.com/cilium/cilium/pull/17592
- [x] https://github.com/cilium/cilium/pull/17593
- [x] https://github.com/cilium/cilium/pull/17595
- [x] https://github.com/cilium/cilium/pull/17596
- [x] https://github.com/cilium/cilium/pull/17720
- [x] https://github.com/cilium/cilium/pull/17718
- [ ] https://github.com/cilium/cilium/pull/17758
- [ ] https://github.com/cilium/cilium/pull/17637

Add alpha-level support for arbitrary configurations for the Cilium Envoy host proxy. The new CiliumEnvoyConfig CRD supports listeners, routes, clusters, and "cluster endpoint sets" (ClusterLoadAssignment) xDS resource types. In addition a listener configured in a CRD can be bound to k8s services, causing load balancing of the services to be postponed to the Envoy listener. The listener must be configured to route to the service backends via a clusters named as "_namespace_/_service_". Cilium agent will automatically push a simple EDS resources with the service backends matching those names. The CRD can use any other load balancing strategy via configuring a Cluster and ClusterLoadAssignment resources using some other name.

Cilium agent will automatically fill in the listener address and xDS config source information, if missing.

Limitations of this release include:
- Listeners on pod ingress will not work yet (not tested)
- For egress, normal bpf datapath policy enforcement is performed after the L7 LB proxy by tail calling from bpf_host to source bpf_lxc. This has the effect where the Envoy proxy will return a 503 reply if L3 to a backend is dropped.
- There is no K8s level validation on the Envoy resources in the new CRD. Any errors will only be visible in Cilium agent log error/warning messages.

Datapath implementation adds a new flag to the main LB map value that causes LB to be bypassed in the socket layer, and the traffic to be redirected to the Envoy listener port on the TC layer. Redirection is done using the Linux transparent proxy mechanism so the Envoy listener receives the traffic with the original destination address (e.g., the ClusterIP of the k8s service).

The new datapath flag is visible in `cilium bpf lb list` command output as `l7-load-balancer`.

A new proxy magic mark value is used for upstream connections from the L7 LB proxy. This mark includes the source endpoint ID so that the datapath can tail call to the source pod's bpf egress path for policy enforcement.

A new CT table value flag `from_l7lb` is added to allow reply packets to be forwarded back to the L7 LB proxy.

CEC CRD can specify `backendServices` causing backends of those services to be synced with Envoy. This is useful for configuring an k8s-level Ingress service that load balances to backends of other services.

A CEC CRD listener will be reachable on a NodePort if `services` contains a service with a node port. This is useful for configuring an k8s-level Ingress service, e.g., of type `LoadBalancer`.

**Note:** Currently service redirected to L7 LB must have backends for Cilium Agent's service cache to wire the service to eBPF datapath. L7 LB continues to work even after such backends are removed. Best practice for now for K8s-level ingress service is to define a dummy backend in some IP address and never use it as a destination in the Listener route configuration. Future versions will lift this requirement.

**Note:** CEC CRD service access from a local node host namespace requires Linux kernel 5.12 or newer, and may not work properly in Kind.

**Note:** Service backends see traffic from external sources via Node Port as coming from the local or remote node (`node`/`remote-node`). This is not ideal for policy enforcement.

Examples of the new CRDs are available at https://github.com/cilium/cilium-cli/pull/550 (see [this PR description](https://github.com/cilium/cilium-cli/pull/550) for demo instructions):
- Simple prometheus listener for the host port 9090: [`connectivity/manifests/envoy-prometheus-metrics-listener.yaml`](https://github.com/cilium/cilium-cli/blob/1df8e05f06f5b88165d1922502a896605f8f1aab/connectivity/manifests/envoy-prometheus-metrics-listener.yaml)
- Prometheus listener using multiple xDS resources: [`connectivity/manifests/envoy-prometheus-metrics-listener-multiresource.yaml`](https://github.com/cilium/cilium-cli/blob/1df8e05f06f5b88165d1922502a896605f8f1aab/connectivity/manifests/envoy-prometheus-metrics-listener-multiresource.yaml)
- L7 LB CRD for the `cilium-test/echo-same-node` and `cilium-test/echo-other-node` services deployed by `cilium connectivity test`: [`connectivity/manifests/envoy-test.yaml`](https://github.com/cilium/cilium-cli/blob/1df8e05f06f5b88165d1922502a896605f8f1aab/connectivity/manifests/envoy-test.yaml)
  - HTTP paths starting with `/foo` is always rewritten to `/`
  - Both services are 50/50 load balanced to the backends of both services
- K8s Ingress service `my-ingress:80` and corresponding CEC CRD building on top of the previous example: [`connectivity/manifests/envoy-ingress-test.yaml`](https://github.com/cilium/cilium-cli/blob/1df8e05f06f5b88165d1922502a896605f8f1aab/connectivity/manifests/envoy-ingress-test.yaml)

```release-note
Initial support for CiliumEnvoyConfig CRD with L7 load balancing is added.
```

jrajahalme committed:
k8s: Add CiliumEnvoyConfig CRD

Add a new CiliumEnvoyConfig to cilium.io API v2alpha1

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
envoy: Switch to protobuf Go APIv2

Switch to protobuf Go APIv2 by changing imports from
'github.com/golang/protobuf' to 'google.golang.org/protobuf'.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
envoy: Call 'callback' also when using current resource version

Call the completion callback also when there is no change in the
cached resource and current version is used, but it has not been
acknowledged yet.

In current use the prior behavior did not cause a problem, but the new
Envoy CRD code depends on the callback to be called also in this case.

Note that the callback is NOT called when there is
no change in the resource and the current version has already been
acknowledged.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
envoy: Add xDS support for RDS, CDS, & EDS

Add support for Route, Cluster, and Endpoint discovery services.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
lb: Add support for L7 load balancers

Add new SVC_FLAG_L7LOADBALANCER to designate services that are not
loadbalanced in BPF, but are forwarded to a local TPROXY loadbalancer
instead.

The set of proxy ports registered to datapath was limited to the
preconfigured set of HTTP, DNS, and proxylib ports. Allow adding (CRD)
proxy ports to this set.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

errordeveloper reviewed (commented):
Just from reading commit messages it seems like quite a few commits could be split into a separate PR that can be merged ahead of this one, to it looks that would at least half of the commit that are currently included here.

errordeveloper commented on pkg/k8s/apis/cilium.io/v2/cec_types.go:
```diff
@@ -0,0 +1,115 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2021 Authors of Cilium
+
+package v2
```
It's a new API, why is it v2 already?

jrajahalme replied:
I had serious trouble with `v2alpha1` but it turned out to be an issue with the only existing `v2alpha1` API I was patterning my work on.. will switch back to `v2alpha1` later, assuming that would be the right thing to do.

jrajahalme replied:
Done.

jrajahalme committed:
service: Push Envoy Endpoints for L7 LB k8s services

Upsert Envoy endpoint resources for services that have been selected
for L7 LB in CiliumEnvoyConfig CRDs.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
envoy: Add xDS resource validation

Make sure Envoy xDS resources have names and validate xDS resources
before adding them to xDS cache. This validation can not be done at
the k8s CRD apply time, so k8s CRD can contain invalid Envoy
configurations.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
envoy: Allocate CRD listener ports if unspecified

Add support for dynamically created (CRD) proxy ports in pkg/proxy.

Allocate proxy ports for CRD listeners without Listener.Address
specification and install datapath TPROXY rules if
Listener.Address is left empty.

Allocated proxy ports are sticky in order to return the same port
when the same listener (by name) is reconfigured or configured again
after deletion.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
envoy: Register k8s service redirection to CRD listener

Register L7 LB service redirections. Use 'L7LBInfo' to keep track of
the CiliumEnvoyConfig that requested a service to be registered for L7
load balancing. Only one CiliumEnvoyConfig may request redirection of
a given service at a time, as otherwise the choice between Envoy
listeners would be ambiguous.

Service backends are synced to Envoy, so that the ingress service can
load balance to them. Datapath L7LB redirection is only requested for
'services' in a CiliumEnvoyResources. Backends are synced to Envoy
also for 'backendServices'. This allows the regular eBPF LB to be used
for pod-to-pod traffic for 'backendServices' while L7 LB is used for
ingress to those same services.

Note: 'Ingress' CEC configurations are still experimental. 

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

christarazi commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -113,6 +113,9 @@ func (ipt *ipt) getVersion() (semver.Version, error) {
 func (ipt *ipt) runProgCombinedOutput(args []string, quiet bool) ([]byte, error) {
+	if args[0] != "-w" {
+		return nil, fmt.Errorf("Missing iptables wait arg (-w): %v", args)
+	}
```
While the code now includes the `-w` flag as the first arg to `iptables`, this check wouldn't suffice in the case where future invocations of `iptables` might not include `-w` as the first arg, i.e. someone adds a new invocation somewhere and isn't aware of this. I think if we still believe this check here is useful to have, then I'd suggest making it a for loop instead to find the `-w`. Or you can make `runProgCombinedOutput` automatically inject `-w` without the user being aware. My opinion / immediate thought is to probably replace this logic with a comment on `runProgCombinedOutput` instead. That'll do the job of informing the user of this function to add the`-w` flag.

Also, if you decide to keep it, then the [error strings should not be capitalized](https://github.com/golang/go/wiki/CodeReviewComments#error-strings).

jrajahalme replied:
Followed your suggestion and made `runProgCombinedOutput()` always inject the wait args, removing this burden from everywhere else.

jrajahalme replied:
Separated this iptables change to a PR of it's own: #17593

christarazi commented on bpf/bpf_lxc.c:
```diff
@@ -1069,7 +1069,7 @@ ipv6_policy(struct __ctx_buff *ctx, int ifindex, __u32 src_label, __u8 *reason,
 		 */
 		send_trace_notify6(ctx, TRACE_TO_PROXY, src_label, SECLABEL, &orig_sip,
-				  0, ifindex, 0, monitor);
+				   0, ifindex, 0, monitor);
```
I assume this is unintended

christarazi replied:
Actually, I notice most changes are prepending an extra space?

joestringer replied:
FWIW These extra spaces seem to be right, the args weren't correctly aligned with the open bracket on the previous line. Could probably be a separate commit but it's not such a big deal given that this commit only has ~25 LoC changes.

jrajahalme replied:
Extra space here and elsewhere is to fix the indentation to match the previous line. But wait, there is more...

This was one of the `TRACE_TO_PROXY` sites that were in consideration for getting the proxy port as a parameter. This, however, is for return traffic for which the proxy port is not used. Nonetheless, the indentation was wrong, so I decided to fix it, while also documenting that I did consider this as a candidate for change as well.

christarazi commented on pkg/monitor/datapath_trace.go:
```diff
@@ -156,7 +156,11 @@ func (n *TraceNotify) traceSummary() string {
+		if n.DstID != 0 {
+			pp = fmt.Sprintf(" port %d", n.DstID)
+		}
+		return fmt.Sprintf("-> proxy%s", pp)
```
My gut feeling is that `Sprintf` will be slower than a simple string concat. operation. Let's change this to

```
return "-> proxy" + pp
```

jrajahalme replied:
Done. Also separated this change to a PR of its own: #17595

christarazi commented on pkg/envoy/json.go:
```diff
@@ -0,0 +1,146 @@
+	envoy_config_listener "github.com/cilium/proxy/go/envoy/config/listener/v3"
+)
+
+// Resources
```
Incomplete?

christarazi replied:
Would be good to add godoc to all exported types

jrajahalme replied:
Done.

christarazi commented on pkg/k8s/apis/cilium.io/v2alpha1/cec_types.go:
```diff
@@ -0,0 +1,141 @@
+	"github.com/cilium/cilium/pkg/option"
+
+	anypb "google.golang.org/protobuf/types/known/anypb"
+
```
Nit: extra newline

jrajahalme replied:
Removed.

christarazi commented on pkg/k8s/watchers/cilium_envoy_config_test.go:
```diff
@@ -0,0 +1,95 @@
+	"sigs.k8s.io/yaml"
+
+	"github.com/cilium/cilium/pkg/envoy"
+	cilium_v2alpha1 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2alpha1"
```
Import ordering

jrajahalme replied:
Fixed.

christarazi commented on pkg/service/service.go:
```diff
@@ -243,6 +253,8 @@ func (s *Service) UpsertService(params *lb.SVC) (bool, lb.ID, error) {
 
 		logfields.LoadBalancerSourceRanges: params.LoadBalancerSourceRanges,
+
+		"l7LBProxyPort": params.L7LBProxyPort,
```
Let's add a corresponding field to the `logfields` pkg and use it here instead

jrajahalme replied:
Done.

christarazi commented on pkg/service/service.go:
```diff
@@ -794,7 +790,6 @@ func (s *Service) upsertServiceIntoLBMaps(svc *svcInfo, onlyLocalBackends bool,
 		SessionAffinityTimeoutSec: svc.sessionAffinityTimeoutSec,
 		CheckSourceRange:          checkLBSrcRange,
 		UseMaglev:                 svc.useMaglev(),
-		L7LBProxyPort:             svc.l7LBProxyPort,
```
Why is this being removed when a previous commit added it? Can we simplify the commits instead?

jrajahalme replied:
Yes, there is a lot to be cleaned with the commits that I did not have the time for just yet.

christarazi commented on pkg/k8s/watchers/cilium_envoy_config.go:
```diff
@@ -0,0 +1,160 @@
+		&cilium_v2alpha1.CiliumEnvoyConfig{},
+		0,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				var valid, equal bool
+				defer func() { k.K8sEventReceived(metricCEC, metricCreate, valid, equal) }()
+				if cec := k8s.ObjToCEC(obj); cec != nil {
+					valid = true
+					err := k.addCiliumEnvoyConfig(cec)
+					k.K8sEventProcessed(metricCEC, metricCreate, err == nil)
+				}
+			},
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				var valid, equal bool
+				defer func() { k.K8sEventReceived(metricCEC, metricUpdate, valid, equal) }()
+
+				if oldCEC := k8s.ObjToCEC(oldObj); oldCEC != nil {
+					if newCEC := k8s.ObjToCEC(newObj); newCEC != nil {
+						valid = true
+						if newCEC.DeepEqual(oldCEC) {
+							equal = true
+							return
+						}
+						err := k.updateCiliumEnvoyConfig(oldCEC, newCEC)
+						k.K8sEventProcessed(metricCEC, metricUpdate, err == nil)
+					}
+				}
+			},
+			DeleteFunc: func(obj interface{}) {
+				var valid, equal bool
+				defer func() { k.K8sEventReceived(metricCEC, metricDelete, valid, equal) }()
+				cec := k8s.ObjToCEC(obj)
+				if cec == nil {
+					return
+				}
+				valid = true
+				err := k.deleteCiliumEnvoyConfig(cec)
+				k.K8sEventProcessed(metricCEC, metricDelete, err == nil)
+			},
```
We have an open issue to refactor any new additions to the k8s watcher to use the `subscriber` pattern. See https://github.com/cilium/cilium/issues/15471. The issue gives an example to follow. One suggestion: the intention of the `subscriber` package is to be extended whenever there's a new resource to be notified for. This PR adds the CEC object, so you can make your changes to `pkg/k8s/watchers/subscriber/cec.go` (needs to be created). Let me know if I can clarify / help with this.

jrajahalme replied:
I'll take a stab at this, was not aware of [#15471](https://github.com/cilium/cilium/issues/15471).

jrajahalme replied:
I tried to find example use or `subscriber.NewRaw()` from the repo and could not. Could you point to an example how to use it for implementing a k8s resource watcher?

christarazi commented on pkg/envoy/json_test.go:
```diff
@@ -0,0 +1,144 @@
+	. "gopkg.in/check.v1"
+)
+
+type JSONSuite struct{}
```
We have an open issue to move to the standard Go testing framework, rather than to add new tests using the old go-check framework: https://github.com/cilium/cilium/issues/16860

jrajahalme replied:
Would be nice to have a conversion guide, i.e., how to do comparable asserts for `IsNil`, `HasLen`, `Equals` (for simpler and deeper types). Until we have that I'm inclined not to invest the time right now on converting the tests.

christarazi replied:
Fair enough. FWIW, the tests would be converted to use testify/assert which has a pretty straightforward set of helpers to do assertions: https://pkg.go.dev/github.com/stretchr/testify/assert

jrajahalme replied:
Based on what I read about `testify` I learned it uses `go-spew` that has an open bug for three years now regarding loops in data (protobuf generated) structures leading to infinite recursion. First, I've had to deal with similar problems in the go-check framework, and would rather not revisit. Second, `testify` does seem to come with its own share of baggage.

christarazi reviewed (changes requested):
Amazing! This is my first pass through. I didn't do a deep inspection of the code and more just took an inventory of what's changing at a high-level. I'll make another pass at some point.

qmonnet commented on Documentation/cmdref/cilium-bugtool_completion_zsh.md:
```diff
@@ -0,0 +1,40 @@
+To load completions for every new session, execute once:
+# Linux:
+$ cilium-bugtool completion zsh > "${fpath[1]}/_cilium-bugtool"
+# macOS:
```
On a quick look at the new documentation files, it looks like Markdown for the description of the completion commands will be broken.
I understand these are generated files, do you know if we have a way to format or include this text differently somehow?

jrajahalme replied:
I don't know, will need to look into this, unless someone else already knows what to do?

jrajahalme committed:
bpf: Add support for re-entering LXC egress path after L7 LB

Allow host to tail call to lxc egress for policy enforcement after L7 LB.

This requires L7 LB proxy to set the source Endpoint ID in the socket
mark. This mark includes a new magic value that the datapath
recognizes to subject the traffic to source LXC (policy enforcement)
processing again. Note that this does not loop endlessly due to the L7
LB switching from service frontend IP to backend IP.

iptables rules are amended to avoid overwriting the new "from L7 LB"
mark with "from host" mark. L7 load balancer upstream connections may
go to the egress proxy listener of the source pod. Pod connections to
proxy listeners are already not tracked to avoid conflicts due to
reuse of the same 5-tuple on both sides of the (egress) proxy. Add
NOTRACK rules for L7 LB connections that may go to the egress proxy
for the same reason.

A new flag 'from_l7lb' is added to conntrack to enable redirecting
reply packets back to L7 LB proxy.

Add a new map for bpf_host to tail call into for LXC policy
enforcement after L7 LB.

Some kernels reached their eBPF complexity limit with the new code
compiled in. Address this by making all L7 LB related code in
bpf_lxc.c confitional on the new ENABLE_L7_LB. This allows the new
feature to be used on kernels that allow a but more complex eBPF
programs.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

joamaki commented on bpf/bpf_host.c:
```diff
@@ -844,6 +844,15 @@ do_netdev(struct __ctx_buff *ctx, __u16 proto, const bool from_host)
 	__u32 __maybe_unused ipcache_srcid = 0;
 	int ret;
 
+	if (from_host) {
```
Just a thought: there's quite a few `if (from_host)` branches here. Might make sense to collapse them and perhaps move the bodies into inlined functions?

joamaki replied:
Also it might sense to put this feature behind an `#ifdef` if possible as it does seem to add quite a few branches to the load-balancer hot path.

jrajahalme replied:
Adding `#ifdef`, but will leave more far reaching refactoring for later.

joamaki replied:
Thanks!

joamaki commented on bpf/bpf_lxc.c:
```diff
@@ -1644,6 +1681,61 @@ int handle_policy(struct __ctx_buff *ctx)
+	switch (proto) {
+#ifdef ENABLE_IPV6
+	case bpf_htons(ETH_P_IPV6):
+		invoke_tailcall_if(__or(__and(is_defined(ENABLE_IPV4), is_defined(ENABLE_IPV6)),
```
why do we only invoke it when both IPv4 and IPv6 are enabled? Also we know for sure IPv6 is enabled here.

jrajahalme replied:
that is some tail_call/complexity optimization logic I have copy/pasted from elsewhere in the file. My reading is that the function will be tail-called into ONLY if both IPv4 and IPv6 are defined, or DEBUG is defined, otherwise the function will be inlined (without a tail call).

IMO we should keep this form intact and not optimize it for the current context (where ENABLE_IPV6 is defined).

joamaki replied:
Ah I see, thanks!

joamaki commented on bpf/bpf_lxc.c:
```diff
@@ -1644,6 +1681,61 @@ int handle_policy(struct __ctx_buff *ctx)
+#endif /* ENABLE_IPV6 */
+#ifdef ENABLE_IPV4
+	case bpf_htons(ETH_P_IP):
+		invoke_tailcall_if(__or(__and(is_defined(ENABLE_IPV4), is_defined(ENABLE_IPV6)),
```
same here as above

joamaki commented on bpf/lib/lb.h:
```diff
@@ -324,6 +324,18 @@ bool lb4_svc_is_localredirect(const struct lb4_service *svc __maybe_unused)
 }
 
+static __always_inline
+bool lb4_svc_is_l7loadbalancer(const struct lb4_service *svc __maybe_unused)
```
drop `__maybe_unused`

jrajahalme replied:
Done.

joamaki commented on bpf/lib/lb.h:
```diff
@@ -324,6 +324,18 @@ bool lb4_svc_is_localredirect(const struct lb4_service *svc __maybe_unused)
+}
+
+static __always_inline
+bool lb6_svc_is_l7loadbalancer(const struct lb6_service *svc __maybe_unused)
```
drop `__maybe_unused`

jrajahalme replied:
Will do (had copy pasted it from above)

jrajahalme replied:
Done.

joamaki commented on pkg/datapath/linux/linux_defaults/mark.go:
```diff
@@ -46,6 +50,18 @@ const (
+	// proxy which is capturing traffic before it is subject to ingress
+	// policy enforcement that must be done after the proxy. The identity
+	// stored in the mark is destination Endpoint ID.
+	MagicMarkIngressEPID int = 0x0800
```
Does this not overlap with `MagicMarkK8sDrop`? Does that matter?

jrajahalme replied:
No, this is in a different nibble (`0x8000`). All four bits covered by `0x0F00` are used by Cilium.

jrajahalme committed:
envoy: Clarify xDS revert function locking

xDS revert functions can not be called from the callback functions
provided to xDS cache update functions due to a locking that would
lead to a deadlock. Revert functions are designed to be called after a
larger transaction fails, not after a single resource update
fails. This is why the revert functions assume to be called outside of
any given callback function and therefore take the locks needed for
that.

Clarify this and add a type that calls the revert functions n reverse
order.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
envoy: Do not wait for RDS, CDS, EDS, or SDS

Do not wait for routes (RDS), clusters (CDS), cluster load assignments
(EDS), or secrets (SDS) as they will only be acked if referenced from
a listener that may not exist yet or may have been deleted already.

Add Cilium agent option --envoy-config-timeout to limit the maximum
time the agent waits for Envoy (Listener) configuration to be acked,
defaulting to 2 minutes. This unblocks k8s watchers if the agent ever
waits for Envoy for resources that are not being acknoowledged.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
datapath: Redirect nodeport L7 LB services

Redirect L7 LB nodeport service connections received by the host to an
Envoy listener. Prior to this commit these services worked as intended
only when the service connection was originated by a pod. This commit
implements L7 LB redirection for nodeport services received by the
node from other nodes or from an external load balancer.

Redirect to nodeport L7 LB from bpf_sock for hostns sources. This is
needed due to this case not being handled by eBPF datapath on TC
level. Note that this requires a recent kernel to work (Linux
5.12+). With older kernels nodeport services only work between nodes,
not from a node host namespace to the node itself.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
envoy: Initial SDS support

Add Initial support for Secret Discovery Service. Cilium xds cluster
details need to be given on each instance, omitting SDS config source
is not yet supported.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

errordeveloper reviewed (commented):
I wonder if the Envoy proxies absolutely must run in the host network namespace? I do also think that at some stage the proxies will need to be managed outside of Cilium pod, but presently, it'd make sense to at least outline the reasoning for why proxies are not running in their own network namespaces. Supposedly, use of host network is just a historic assumption, but that still would be good to confirm.
One would think that host port management is not very trivial sort of business, especially as ports can be a scarce resource. On the other hand, having the Envoy proxies exposed to the host network by default also seems like a security concern.
Other than that, not much has come to my attention while reading the code. I can see that a few reviewers had looked at the changes line by line already.

jrajahalme committed:
datapath: Reduce from LXC complexity

Split from LXC code paths into two tail calls when per packet load
balancing is needed. When per packet load balancing is not needed this
should have minimal impact on datapath performance.

IPv6 code path is refactored to pattern it more similar to IPv4 code
path.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
Envoy: Update API and image for CiliumEnvoyConfig support

Envoy Go API is updated to contain the generated validation code.

Envoy image is updated to support the new EndpointId option for the
bpf_metadata listener filter. NPDS field 'Policy' is renamed as
'EndpointID'. 'Policy' field was not used for anything, so might as
well recycle it while this API is not yet public.

Envoy retries may fail on "address already in use" when the original
source address and port are used on upstream connections. Cilium
typically does this in the egress proxy listeners. Fix this by using a
Cilium Envoy build that always sets SO_REUSEADDR when original source
address and port is used.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>
